x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatePeriod.java

```diff
@@ -0,0 +1,156 @@
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.type.EsqlDataTypeConverter;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.esql.core.type.DataType.DATE_PERIOD;
+import static org.elasticsearch.xpack.esql.core.type.DataType.KEYWORD;
+import static org.elasticsearch.xpack.esql.core.type.DataType.TEXT;
+import static org.elasticsearch.xpack.esql.core.type.DataType.isString;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypeConverter.DATE_PERIODS;
+
+public class ToDatePeriod extends AbstractConvertFunction {
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(
+        Expression.class,
+        "ToDatePeriod",
+        ToDatePeriod::new
+    );
+
+    private static final Map<DataType, BuildFactory> EVALUATORS = Map.ofEntries(
+        Map.entry(DATE_PERIOD, (fieldEval, source) -> fieldEval),
+        Map.entry(KEYWORD, (fieldEval, source) -> fieldEval),
+        Map.entry(TEXT, (fieldEval, source) -> fieldEval)
+    );
+
+    private static final String INVALID_INTERVAL_ERROR = "Invalid interval value in [{}], expected integer followed by one of "
+        + DATE_PERIODS
+        + " but got [{}]";
+
+    @FunctionInfo(
+        returnType = "date_period",
+        description = "Converts a string into a `date_period` value.",
+        examples = @Example(file = "convert", tag = "castToDatePeriod")
+    )
+    public ToDatePeriod(
+        Source source,
+        @Param(name = "string", type = { "date_period", "keyword", "text" }, description = "A string.") Expression v
+    ) {
+        super(source, v);
+    }
+
+    private ToDatePeriod(StreamInput in) throws IOException {
+        super(in);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    protected Map<DataType, BuildFactory> factories() {
+        return EVALUATORS;
+    }
+
+    @Override
+    protected final TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+        TypeResolution typeResolution = isType(
+            field(),
+            factories()::containsKey,
+            sourceText(),
+            FIRST,
+            supportedTypesNames(supportedTypes())
+        );
+        if (field().foldable()) {
+            if (isString(field().dataType()) && isValidInterval(field().fold().toString()) == false) {
+                typeResolution = new TypeResolution(LoggerMessageFormat.format(null, INVALID_INTERVAL_ERROR, sourceText(), field().fold()));
+            }
+
+        }
```

Comment:
nit: similarly to `Percentile`, we could probably just require this to be foldable here. `Percentile`'s `resolveType` uses `.and(isFoldable(...))` which we could re-use here (+ keep verifying the validity of the interval).

Reply:
I was trying to make cases like below work, in this case the input to `::time_duration` is a `ReferenceAttribute`, and it is not foldable, if only foldable expressions can pass `Analyzer`,  cases like this will be blocked. The validation of a `ReferenceAttribute` is done in `LogicalPlanOptimizer` after `PropagateEvalFoldables`. Let me see if I can add some restrictions here to make foldable expression like `Concat` work.

```
FROM employees
| EVAL interval = "3 hours", x = birth_date + interval::time_duration
```

Reply:
That would be neat, indeed! If we can make this work, let's add csv tests for this as well.

I recall that for cases like this, we have the `Validatable` interface - implemented e.g. by `Bucket`, which also requries its 2nd - 4th arguments to be foldable.

I guess if we used that, `Validatable#validate` would just return a failure when called, because `to_dateperiod` should just be folded away after optimization?

Reply:
Another trick we could try is to just always declare `to_dateperiod` as foldable, and have it fail at constant folding time.

Reply:
Yes, `Validatable` does not apply to here, because these two conversion methods are fold to constant before `Validatable` sees them.